### ISSUE TITLE
[Feature Request] getQueryKey() function to help with optimistic updates in tanstack-Query

### ISSUE THREAD

**Issue by ErikDakoda** [zenstackhq/zenstack#697]
Background: In tanstack-query the query key is required for `queryClient.setQueryData()` to implement optimistic updates.

Currently a query key is a triad of :[ModelName, QueryURL, QueryArgs], so for example for useUserFindUnique, the key will be: ['User', 'http://.../user/findUnique%60, queryArgs]. Computing it manually is not recommended, as it may change in the future.

Here is how this may work:

```js
import { getQueryKey } from "@zenstackhq/tanstack-query";

const queryClient = useQueryClient();

// get query key for the current post
const postKey = getQueryKey('Post', 'findUnique', { where: { id: postId } });

const updatePost = useUpdatePost({
  onMutate: (newData) => {
    const previous = queryClient.getQueryData(postKey);
    // optimistically update the current post
    queryClient.setQueryData(postKey, (old) => ({...old, ...newData.data});
    return { previous };
  },
  onError: (err, newData, context) => {
    queryClient.setQueryData(postKey, context.previous);
  }
});

const onUpdate = () => {
  updatePost.mutate(...);
};
```

It would be great to also have a `getQueryKeys()` function that accepts partial parameters and returns an array of matches. For example `getQueryKeys({ modelName: 'Post' })` for all Post queries or `getQueryKeys({ queryArgs: { select: { posts: true } } })` for all queries with a relation to posts.

It would also be nice if the result object returned by query functions like `findUnique` would also include the query key. Of course, if you are just passing the result returned by tanstack-query you may not want to mutate that object.

**Comment by Azzerty23** [comment]
I've just seen that React Query v5 has re-implemented optimistic updates. As I understand it, we don't need to retrieve the old cached data anymore. Instead `useMutation()` hook returns `variables` directly, and if we want to retrieve these variables in another place, we can do so via `useMutationState()` by passing the `mutationKey`. So, we shouldn't need a function like `getQueryKey()` I guess.

Waiting https://github.com/zenstackhq/zenstack/issues/780 from now 😄

**Comment by Azzerty23** [comment]
Sorry, still needed for invalidate cache

**Comment by ymc9** [comment]
`getQueryKey` function is now available in v1.2.0 through `@zenstackhq/tanstack-query/runtime` and `@zenstackhq/swr/runtime`.

Closing this issue for now. Automatic optimistic update is tracked separately here: #795
